src/main/scala/io/iohk/ethereum/network/p2p/MessageCodec.scala

```diff
@@ -0,0 +1,38 @@
+package io.iohk.ethereum.network.p2p
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.util.ByteString
+import io.iohk.ethereum.rlp
+import io.iohk.ethereum.rlp.RLPEncoder
+
+import scala.util.Try
+
+class MessageCodec(frameCodec: FrameCodec) {
+
+  val MaxFramePayloadSize: Int = Int.MaxValue // no framing
+
+  val contextIdCounter = new AtomicInteger
+
+  def readMessages(data: ByteString): Seq[Try[Message]] = {
+    val frames = frameCodec.readFrames(data)
+    // TODO: protocolVersion?
+    frames map { frame => Try(Message.decode(frame.`type`, frame.payload.toArray, Message.PV61)) }
+  }
+
+  def encodeMessage[M <: Message : RLPEncoder](message: M): ByteString = {
+    val encoded = rlp.encode(message)
+    val numFrames = Math.ceil(encoded.length / MaxFramePayloadSize.toDouble).toInt
+    val contextId = contextIdCounter.incrementAndGet()
+
+    val frames = (0 until numFrames) map { frameNo =>
+      val payload = encoded.drop(frameNo * MaxFramePayloadSize).take(MaxFramePayloadSize)
```

Comment:
Minor detail: isn't it more efficient to use the slice function instead of drop and take as with: `encoded.slice(frameNo * MaxFramePayloadSize, frameNo * MaxFramePayloadSize + MaxFramePayloadSize)`?